cuda/solver/lower_trs_kernels.cu

```diff
@@ -53,10 +66,96 @@ namespace cuda {
 namespace lower_trs {
 
 
+void should_perform_transpose(std::shared_ptr<const CudaExecutor> exec,
+                              bool &do_transpose)
+{
+#if (defined(CUDA_VERSION) && (CUDA_VERSION >= 9020))
+
+
+    do_transpose = false;
+
+
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+
+
+    do_transpose = true;
+
+
+#endif
+}
+
+
+void init_struct(std::shared_ptr<const CudaExecutor> exec,
+                 std::shared_ptr<gko::solver::SolveStruct> &solve_struct)
+{
+    solve_struct = std::shared_ptr<gko::solver::SolveStruct>(
+        new gko::solver::SolveStruct());
+}
+
+
 template <typename ValueType, typename IndexType>
 void generate(std::shared_ptr<const CudaExecutor> exec,
               const matrix::Csr<ValueType, IndexType> *matrix,
-              const matrix::Dense<ValueType> *b) GKO_NOT_IMPLEMENTED;
+              gko::solver::SolveStruct *solve_struct,
+              const gko::size_type num_rhs)
+{
+    if (cusparse::is_supported<ValueType, IndexType>::value) {
+        auto handle = exec->get_cusparse_handle();
+
+
+#if (defined(CUDA_VERSION) && (CUDA_VERSION >= 9020))
+
+
+        ValueType one = 1.0;
+
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseSetPointerMode(handle, CUSPARSE_POINTER_MODE_HOST));
+        cusparse::buffer_size_ext(
+            handle, solve_struct->algorithm, CUSPARSE_OPERATION_NON_TRANSPOSE,
+            CUSPARSE_OPERATION_TRANSPOSE, matrix->get_size()[0], num_rhs,
+            matrix->get_num_stored_elements(), &one, solve_struct->factor_descr,
+            matrix->get_const_values(), matrix->get_const_row_ptrs(),
+            matrix->get_const_col_idxs(), nullptr, num_rhs,
+            solve_struct->solve_info, solve_struct->policy,
+            &solve_struct->factor_work_size);
+
+        // allocate workspace
+        solve_struct->factor_work_vec =
+            exec->alloc<void *>(solve_struct->factor_work_size);
```

Comment:
Is it possible that allocate the memory before free it?

Reply:
Do you mean an adding and additional check like this ? 

```c++
        if (solve_struct->factor_work_vec != nullptr) {
            GKO_ASSERT_NO_CUDA_ERRORS(cudaFree(solve_struct->factor_work_vec));
        }

```